ENH: Allow map with abc mapping

ohad83 opened the pull request:
- [x] closes #29733 (partly, look at my issue comment for details on why it won't exactly work on the issue's example)
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

ohad83 committed:
BUG - Allow a mapper of type collections.abc.Mapping when using
Series.map

ohad83 committed:
BUG - Fixed Series.map to work with no-missing ABC Mapping

ohad83 committed:
TST - Added a test for Series.map with dict-like (abc.Mapping) mapper

ohad83 committed:
BUG - In Series.map, only look for __missing__ if the mapper is a dict
subclass.
TST - Add a test for the aforementioned bug

ohad83 committed:
CLN - Add a newline for readability

ohad83 committed:
CLN - Import sorting fix

ohad83 committed:
DOC - Added collections.abc.Mapping to arg doc of Series.map
DOC - Moved whatsnew of the aforementioned change from BUG to ENH

ohad83 commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -559,6 +559,7 @@ Other
 - Bug in :meth:`DataFrame.append` that raised ``IndexError`` when appending with empty list (:issue:`28769`)
 - Fix :class:`AbstractHolidayCalendar` to return correct results for
   years after 2030 (now goes up to 2200) (:issue:`27790`)
+- Bug in :meth:`Series.map` where the mapper could not be a dict-like object which subclasses `collections.abc.Mapping` (:issue:`29733`)
```
Alright, moved it to enhancement.

jreback commented on pandas/core/series.py:
```diff
@@ -253,7 +254,7 @@ def __init__(
                     data = data.reindex(index, copy=copy)
                 data = data._data
-            elif isinstance(data, dict):
+            elif isinstance(data, abc.Mapping):
```
this is an orthogonal change pls remove

ohad83 replied:
It is not orthogonal.
In the function `_map_values` in `core/base.py`, in case the mapper is a subclass of `collections.abc.Mapping` and isn't a `dict` with the `__missing__` property, we create a Series out of the mapper. In case it is `collections.abc.Mapping` and isn't a `dict`, the creation of the Series will fail if not for this change.

WillAyd replied:
Is this hit in a test case?

ohad83 replied:
It is in the test cases I added. If I remove this change, they will fail, as a Series is created from the mapper.

jreback commented on pandas/core/series.py:
```diff
@@ -3617,7 +3618,7 @@ def map(self, arg, na_action=None):
         Parameters
         ----------
-        arg : function, dict, or Series
+        arg : function, dict, colletions.abc.Mapping subclass or Series
```
dict is now superfluous

ohad83 replied:
Wasn't sure if I should remove it as I suppose many standard users are familiar with `dict` but unfamiliar with `Mapping`.
Removed it now.

jreback commented on pandas/tests/series/test_apply.py:
```diff
@@ -609,6 +609,51 @@ class DictWithoutMissing(dict):
         tm.assert_series_equal(result, expected)
 
+    def test_map_abc_mapping(self):
+        class NonDictMapping(abc.Mapping):
```
add the issue number as a comment

jreback commented on pandas/core/base.py:
```diff
@@ -1215,8 +1215,8 @@ def _map_values(self, mapper, na_action=None):
-        if isinstance(mapper, dict):
-            if hasattr(mapper, "__missing__"):
+        if isinstance(mapper, abc.Mapping):
+            if isinstance(mapper, dict) and hasattr(mapper, "__missing__"):
```
why do you think you need the if condition that was added?

ohad83 replied:
`collections.abc.Mapping` does not declare nor implement `__missing__`, that is a concept of `dict`. In case you use a subclass of `collections.abc.Mapping` which isn't a subclass of `dict`, you cannot rely on `__missing__`.
You can see an example on the second test I added, `test_map_abc_mapping_with_missing`.

WillAyd replied:
I think can just simplify to `isinstance(mapper, dict)` then no? Check for __missing__ after that is duplicative?

ohad83 replied:
No, since `dict` doesn't have a `__missing__` method on its own.
In `dict.__getitem__`'s implementation there is a reference to `__missing__` in case the object is a subclass of `dict` and implements it, but `dict` itself does not implement `__missing__`.
I suggest reading [dict's documentation](https://docs.python.org/3/library/stdtypes.html#dict), specifically the `__getitem__` part noted as `d[key]` for explanation.

ohad83 committed:
DOC - Removed superfluous 'dict' from Series.map argument description
DOC - Added issue number for tests of Series.map with abc.Mapping argument

WillAyd reviewed (changes requested):
Looks good generally. cc @TomAugspurger and @simonjayhawkins for related discussions that we've had on types - not sure if this is something we would need to review holistically

TomAugspurger commented:
Changes here look sensible.

It's not immediately clear to my why `Series.__init__` had to change. Can
we add a test in pandas/tests/series/test_constructors.py that shows the
change?

We may want to move these Mapping subclass test helpers to a central place
and reuse them.



On Fri, Nov 22, 2019 at 12:41 PM William Ayd <notifications@github.com>
wrote:

> *@WillAyd* requested changes on this pull request.
>
> Looks good generally. cc @TomAugspurger <https://github.com/TomAugspurger>
> and @simonjayhawkins <https://github.com/simonjayhawkins> for related
> discussions that we've had on types - not sure if this is something we
> would need to review holistically
> ------------------------------
>
> In pandas/core/base.py
> <https://github.com/pandas-dev/pandas/pull/29788#discussion_r349737661>:
>
> > @@ -1215,8 +1215,8 @@ def _map_values(self, mapper, na_action=None):
>          # we can fastpath dict/Series to an efficient map
>          # as we know that we are not going to have to yield
>          # python types
> -        if isinstance(mapper, dict):
> -            if hasattr(mapper, "__missing__"):
> +        if isinstance(mapper, abc.Mapping):
> +            if isinstance(mapper, dict) and hasattr(mapper, "__missing__"):
>
> I think can just simplify to isinstance(mapper, dict) then no? Check for
> *missing* after that is duplicative?
> ------------------------------
>
> In pandas/core/series.py
> <https://github.com/pandas-dev/pandas/pull/29788#discussion_r349738912>:
>
> > @@ -253,7 +254,7 @@ def __init__(
>                  else:
>                      data = data.reindex(index, copy=copy)
>                  data = data._data
> -            elif isinstance(data, dict):
> +            elif isinstance(data, abc.Mapping):
>
> Is this hit in a test case?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/29788?email_source=notifications&email_token=AAKAOIQADNZIIDNSMG5VKADQVARXPA5CNFSM4JQNDNTKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMWUNVY#pullrequestreview-321734359>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAOIWCCMV5AXPYT46DJITQVARXPANCNFSM4JQNDNTA>
> .
>

ohad83 committed:
TST - Moved NonDictMapping test util class to pandas/util/testing.py
TST - Added a test of constructing a series from NonDictMapping

ohad83 commented:
> Changes here look sensible. It's not immediately clear to my why `Series.__init__` had to change. Can we add a test in pandas/tests/series/test_constructors.py that shows the change? We may want to move these Mapping subclass test helpers to a central place and reuse them.

Added a test for constructing a `Series` with `NonDictMapping` and moved the `NonDictMapping` test subclass to `pandas/util/testing.py`.

simonjayhawkins commented on pandas/core/base.py:
```diff
@@ -2,7 +2,7 @@
 """
 import builtins
-from collections import OrderedDict
+from collections import OrderedDict, abc
```
minor nit. IMO shouldn't import abc from collections. to avoid confusion and potential future clashes with the abc module.

ohad83 replied:
I agree, though there are other files in the code which do that so I figured it's fine.
Changed.

ohad83 replied:
Apparently the guidelines in code_check.sh point to using `from collections import abc` instead of `from collections.abc import`, so changes reverted.

simonjayhawkins replied:
fair enough. not sure of the reason for that.

simonjayhawkins commented on pandas/core/series.py:
```diff
@@ -1,6 +1,7 @@
 """
 Data structure for 1-dimensional cross-sectional and time series data
 """
+from collections import abc
```
same

simonjayhawkins commented on pandas/core/series.py:
```diff
@@ -3617,7 +3618,7 @@ def map(self, arg, na_action=None):
         Parameters
         ----------
-        arg : function, dict, or Series
+        arg : function, colletions.abc.Mapping subclass or Series
```
I think could simply just use "mapping" in docstrings

ohad83 replied:
I am not sure about that, the other docstrings I've seen in pandas use the whole `collections.abc.Mapping` and I believe it's better for the users who aren't familiar with `collections.abc`. Having the whole path can help them search it, while just `Mapping` won't be as helpful.
You did make me notice I had a typo though 😅

ohad83 committed:
CLN - Better way to import from collections.abc
DOC - Typo fix

ohad83 committed:
Revert "CLN - Better way to import from collections.abc"

This reverts commit 5fe41408d7c422734955e2ff74e2e0bdfa897785.
Apparently it's an unwanted pattern checked for by code_checks.sh (from
collections.abc import)

ohad83 committed:
DOC - Type fix

ohad83 committed:
CLN - remove example.feather which was mistakenly added

jreback commented on pandas/core/series.py:
```diff
@@ -253,7 +254,7 @@ def __init__(
                     data = data.reindex(index, copy=copy)
                 data = data._data
-            elif isinstance(data, dict):
+            elif isinstance(data, abc.Mapping):
```
you can use is_dict_like here instead

jreback commented on pandas/core/base.py:
```diff
@@ -1215,8 +1215,8 @@ def _map_values(self, mapper, na_action=None):
         # python types
-        if isinstance(mapper, dict):
-            if hasattr(mapper, "__missing__"):
+        if isinstance(mapper, abc.Mapping):
```
use is_dict_like

ohad83 committed:
CLN: Checking is_dict_like instead of is abc.Mapping

ohad83 committed:
CLN: Changed import sorting

jreback commented:
lgtm. ping on green.

WillAyd reviewed (approved):
lgtm. @ohad83 can you merge master one more time? I think should get CI green

ohad83 commented:
@WillAyd Merged but still a test seems to fail, I am not really sure what's happening and how, if at all, it's related to this PR.. Any ideas?

ohad83 commented:
> lgtm. @ohad83 can you merge master one more time? I think should get CI green

@jreback @WillAyd It's green now

WillAyd commented on pandas/util/testing.py:
```diff
@@ -2121,6 +2121,20 @@ def __init__(self, *args, **kwargs):
         dict.__init__(self, *args, **kwargs)
 
 
+class TestNonDictMapping(abc.Mapping):
```
Can you make this a fixture in `pandas/tests/series/conftest.py` instead of putting in `pandas.util.testing`? With the latter sometimes things can leak into our API and definitely wouldn't want that here

jbrockmendel commented:
@ohad83 can you rebase and address Will's comment if still relevant

jreback commented:
can you merge master and will look again

jreback committed:
use fixtures

jreback commented:
@ohad83 going to have this close the original issue. if the remaining case is worthile, then pls open a new issue.

jreback committed:
removed uneeded import

jreback committed:
moar uneeded imports

jreback committed:
missing import

jreback commented:
@WillAyd if any comments

WillAyd commented:
Thanks @ohad83 great PR
